ci-operator/step-registry/openshift/upgrade/qe/test/paused/openshift-upgrade-qe-test-paused-chain.yaml

```diff
@@ -6,11 +6,12 @@ chain:
   - ref: openshift-extended-upgrade-pre
   - ref: cucushift-upgrade-prepare
   - ref: cucushift-upgrade-prehealthcheck
-  - ref: cucushift-upgrade-paused-intermediate
+  - ref: cucushift-upgrade-paused-pause-worker-mcp
+  - ref: cucushift-chainupgrade-toimage
```

Comment:
`cucushift-chainupgrade-toimage` is only for one-hop upgrade now, and the plan is to enable it for two-hops later, right? but it does not include sanity test, so it looks like we could not just use this chain.

Reply:
> No, cucushift-chainupgrade-toimage is always used as the last hop upgrade, even in the future.

the last hop should be `cucushift-upgrade-toimage` in line 12?

Reply:
Ah, sorry, I miss read the original comment. Let me correct it.

> cucushift-chainupgrade-toimage is only for one-hop upgrade now, and the plan is to enable it for two-hops later, right?

Right.

>  but it does not include sanity test, so it looks like we could not just use this chain.

I think that is a bit out of this PR's topic. 

And I saw the chain already had `openshift-extended-upgrade-pre` and `cucushift-upgrade-prepare` steps before `cucushift-chainupgrade-toimage` to run some e2e testing. 

If you mean we need to run sanity test between every hop upgrade in `cucushift-chainupgrade-toimage`, I have no good idea for that, after all, the current multi-hop non-eus upgrade did not run sanity testing either. 

If we really need to run sanity test between every hop upgrade in `cucushift-chainupgrade-toimage`, then you are right, we can not use this chain. But what is our plan? Copying some step into multiple ones with the same code per the number of the upgrade hop, so that we can have a chance to insert sanity testing?

Reply:
> openshift-extended-upgrade-pre and cucushift-upgrade-prepare

these two steps are for upgrade cases only.

> But what is our plan? Copying some step into multiple ones with the same code per the number of the upgrade hop, so that we can have a chance to insert sanity testing?

The test scope is to do e2e test on each round of semi-upgrade(unmatched master+worker nodes).  That's why there were `cucushift-upgrade-paused-intermediate` and `cucushift-upgrade-paused-latest` splitted for this purpose due to prow limitation itself.  And the same reason that there were `cucushift-upgrade-paused-intermediate-sanity` and `cucushift-upgrade-paused-latest-sanity` with the same.

I'm thinking if we could assemble a new chain for the loop, including both toimage and e2e step. Or else, it sounds like we still need `cucushift-upgrade-paused-intermediate` and `cucushift-upgrade-paused-latest`.

Reply:
If we only take 2 hops eus ugprade into consideration for now, `cucushift-upgrade-paused-intermediate` and `cucushift-upgrade-paused-latest` totally can be replaced by `cucushift-chainupgrade-toimage` and `cucushift-upgrade-toimage`, I do not think we have to maintain them any more.

> I'm thinking if we could assemble a new chain for the loop, including both toimage and e2e step. Or else, it sounds like we still need cucushift-upgrade-paused-intermediate and cucushift-upgrade-paused-latest.

"assemble a new chain for the loop" sounds a potential solution, but I don't think we should copy a step with multiple steps with different names any more, we need some smarter way to do that. Maybe deserve a meeting to discuss it.

Reply:
i'm ok to have a discussion, together with https://github.com/openshift/release/pull/48427#discussion_r1531455585 about `cucushift-chainupgrade-toimage`.

Reply:
per sync result,  `cucushift-chainupgrade-toimage` will be refactor to meet current 3 scenarios(eus, rollback, multi-hop normal upgrade) :

- include sanity test step for eus
- split a dedicated step about dropping last hop or not

Reply:
@jiajliu the code get updated now.

Reply:
That needs a following PR to do (use chainupgrade-toimage in eus and drop upgrade-toimage and e2e sanity testing).

Reply:
sounds good to me